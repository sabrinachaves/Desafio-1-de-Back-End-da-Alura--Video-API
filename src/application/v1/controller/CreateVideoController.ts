import { buildErrorInfo } from '@infrastructure/parser/ErrorInfo';
import ICreateVideoDTO from '@service/CreateVideo/interfaces/ICreateVideoDTO';
import ICreateVideoService from '@service/CreateVideo/interfaces/ICreateVideoService';
import { RequestHandler } from 'express';
import { CREATED } from 'http-status';

export default class CreateVideoController {
  constructor(private createVideoService: ICreateVideoService) {}

  public execute: RequestHandler = async (request, response) => {
    try {
      const { title, description, url } = request.body;

      const createVideoDTO: ICreateVideoDTO = {
        title,
        description,
        url,
      };

      const video = await this.createVideoService.handle(createVideoDTO);

      return response.status(CREATED).json(video);
    } catch (err: any) {
      const errorInfo = buildErrorInfo(err);
      const responseData = { error: errorInfo.errorMessage };

      return response.status(errorInfo.code).json(responseData);
    }
  };
}
