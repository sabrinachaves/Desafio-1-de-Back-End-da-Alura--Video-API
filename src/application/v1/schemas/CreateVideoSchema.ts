import Joi from 'joi';

const createVideoSchema = Joi.object().keys({
  title: Joi.string().required(),
  description: Joi.string(),
  url: Joi.string().required(),
});

export default createVideoSchema;
