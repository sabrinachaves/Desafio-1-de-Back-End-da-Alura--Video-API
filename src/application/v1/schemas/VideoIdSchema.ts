import Joi from 'joi';

const videoIdSchema = Joi.object().keys({
  id: Joi.string().uuid(),
});

export default videoIdSchema;
