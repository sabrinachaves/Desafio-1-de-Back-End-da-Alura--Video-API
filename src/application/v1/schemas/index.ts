import createVideoSchema from './CreateVideoSchema';
import videoIdSchema from './VideoIdSchema';

export default {
  videoId: videoIdSchema,
  createVideo: createVideoSchema,
};
