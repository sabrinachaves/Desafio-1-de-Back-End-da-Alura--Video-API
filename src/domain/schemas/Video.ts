import { Schema } from 'mongoose';
import { v4 as uuidv4 } from 'uuid';

export interface IVideo {
  _id?: string;
  title: string;
  description?: string;
  url: string;
}

export const Video: Schema = new Schema<IVideo>(
  {
    _id: {
      type: String,
      default: uuidv4(),
    },
    title: {
      type: String,
      required: true,
    },
    description: {
      type: String,
    },
    url: {
      type: String,
      required: true,
    },
  },
  { timestamps: true },
);
