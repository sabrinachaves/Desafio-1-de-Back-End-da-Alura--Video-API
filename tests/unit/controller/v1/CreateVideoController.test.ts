import { Request } from 'express';
import { CREATED, UNPROCESSABLE_ENTITY } from 'http-status';
import CreateVideoController from '../../../../src/application/v1/controller/CreateVideoController';
import CreateVideoServiceMock from '../../../mocks/CreateVideoServiceMock';
import { IVideo } from '../../../../src/domain/schemas/Video';
import { videoMock } from '../../../mocks/VideoMock';
import UnprocessableEntityError from '../../../../src/domain/exceptions/UnprocessableEntityError';

describe('CreateVideoController', () => {
  const createVideoService = new CreateVideoServiceMock();
  it('should match the snapshot #unit', () => {
    const instance = new CreateVideoController(createVideoService);
    expect(instance).toMatchSnapshot();
  });

  it('should create the video #unit', async () => {
    createVideoService.handle = jest.fn().mockImplementation((): Promise<IVideo> => {
      return Promise.resolve(videoMock);
    });
    const instance = new CreateVideoController(createVideoService);

    const req = {
      body: {
        title: videoMock.title,
        description: videoMock.description,
        url: videoMock.url,
      },
      get: jest.fn(),
    } as unknown as Request;
    const mockJsonResponse = jest.fn();
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: mockJsonResponse,
      })),
    };

    await instance.execute(req, res, jest.fn());
    expect(res.status).toHaveBeenCalledWith(CREATED);
    expect(mockJsonResponse).toHaveBeenCalledWith(videoMock);
  });

  it('should throw an error when already exists a video with the same url #unit', async () => {
    createVideoService.handle = jest
      .fn()
      .mockRejectedValueOnce(new UnprocessableEntityError(`There's already a video with this url`));

    const instance = new CreateVideoController(createVideoService);

    const req = {
      body: {
        title: videoMock.title,
        description: videoMock.description,
        url: videoMock.url,
      },
      get: jest.fn(),
    } as unknown as Request;
    const mockJsonResponse = jest.fn();
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: mockJsonResponse,
      })),
    };

    await instance.execute(req, res, jest.fn());
    expect(res.status).toHaveBeenCalledWith(UNPROCESSABLE_ENTITY);
    expect(mockJsonResponse).toHaveBeenCalledWith({ error: `There's already a video with this url` });
  });
});
