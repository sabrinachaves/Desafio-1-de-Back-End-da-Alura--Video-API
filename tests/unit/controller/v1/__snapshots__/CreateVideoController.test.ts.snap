// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateVideoController should match the snapshot #unit 1`] = `
CreateVideoController {
  "createVideoService": CreateVideoServiceMock {},
  "execute": [Function],
}
`;
